src/rascal/models/sparse_kernel_predict.hh

```diff
@@ -0,0 +1,256 @@
+
+#include "rascal/math/utils.hh"
+#include "rascal/models/kernels.hh"
+#include "rascal/models/sparse_kernels.hh"
+#include "rascal/structure_managers/property_block_sparse.hh"
+#include "rascal/structure_managers/structure_manager_collection.hh"
+#include "rascal/utils/json_io.hh"
+
+namespace rascal {
+
+  /**
+   * Compute the partial gradients of a structure w.r.t atomic positions
+   * using the GAP model (see
+   * @ref SparseKernelImpl<internal::SparseKernelType::GAP>::compute_derivative)
+   *
+   * The point of this function is to provide a faster prediction routine
+   * compared to computing the kernel element and then multiplying them with
+   * the weights of the model.
+   *
+   * The partial gradients are attached to the input manager in a Property of
+   * Order 2.
+   * partial gradients [N_{neighbor}, 3]
+   *
+   * @tparam StructureManagers should be an iterable over shared pointer
+   *          of structure managers like ManagerCollection
+   * @param sparse_points a SparsePoints* class
+   * @param managers a ManagerCollection or similar collection of
+   * structure managers
+   * @param zeta exponent of the GAP kernel
+   * @param weights of the GAP model
+   * @param representation_name name used to get the representation in the
+   * managers
+   * @param representation_grad_name name used to get representation gradient
+   * the in the managers
+   * @param djfi_name name used to get/register the partial gradients in the
+   * managers
+   * @return
+   */
+  template <class Property_t, class PropertyGradient_t, class StructureManager,
+            class SparsePoints>
+  void
+  compute_partial_gradients_gap(StructureManager & manager,
+                                SparsePoints & sparse_points,
+                                math::Vector_t & weights, const size_t zeta,
+                                const std::string & representation_name,
+                                const std::string & representation_grad_name,
+                                const std::string & djfi_name) {
+    using Manager_t = typename StructureManager::element_type;
+    using Keys_t = typename SparsePoints::Keys_t;
+    using Key_t = typename SparsePoints::Key_t;
+
+    auto && prop{
+        *manager->template get_property<Property_t>(representation_name, true)};
+    auto && prop_grad{*manager->template get_property<PropertyGradient_t>(
+        representation_grad_name, true)};
+
+    const int inner_size{prop.get_nb_comp()};
+
+    bool do_block_by_key_dot{false};
+    if (prop_grad.are_keys_uniform()) {
+      do_block_by_key_dot = true;
+    }
+
+    std::set<int> unique_species{};
+    for (auto center : manager) {
+      unique_species.insert(center.get_atom_type());
+    }
+
+    // find shared central atom species
+    std::set<int> species_intersect{
+        internal::set_intersection(unique_species, sparse_points.species())};
+
+    // find offsets alongs the sparse points direction
```

Comment:
Isn't this (finding offsets) something that should be handled by PropertyBlockSparse?

Reply:
My understanding is that `PropertyBlockSparse` deals with sparse features; while this is dealing with sparse environments. I might be wrong though =)

Reply:
removed because unused